File: scala/leet_code/1700.scala
/**
  * This module provides a solution to the problem of counting the number of students that cannot eat lunch.
  * The students are in a queue and each student either prefers a square sandwich (represented by 0) or a circular sandwich (represented by 1).
  * The sandwiches are in a stack and the sandwich on top of the stack is the one that will be given out next.
  * The students will leave the queue if they get their preferred sandwich, otherwise they will go to the end of the queue.
  * The function `countStudents` is the main function in this module.
  */
object Solution {

  /**
    * Function to count the number of students that cannot eat lunch.
    *
    * @param students an Array[Int] representing the students in the queue. Each student is represented by an integer that indicates their preferred sandwich (0 for square, 1 for circular).
    * @param sandwiches an Array[Int] representing the sandwiches in the stack. Each sandwich is represented by an integer (0 for square, 1 for circular).
    * @return an Int representing the number of students that cannot eat lunch.
    */
  def countStudents(students: Array[Int], sandwiches: Array[Int]): Int = {
    // Initialize a mutable buffer from the students array to represent the queue of students
    val queue = students.toBuffer
    // Initialize a mutable buffer from the sandwiches array to represent the stack of sandwiches
    val stack = sandwiches.toBuffer
    // Initialize a variable to keep track of the number of iterations over the queue
    var i = 0

    // While there are students in the queue and the number of iterations is less than the number of students
    while (i < queue.length && queue.nonEmpty) {
      // If the first student in the queue prefers the sandwich on top of the stack
      if (queue.head == stack.head) {
        // Remove the first student from the queue and the sandwich from the top of the stack
        queue.remove(0)
        stack.remove(0)
        // Reset the number of iterations
        i = 0
      } else {
        // Move the first student to the end of the queue
        queue.append(queue.remove(0))
        // Increment the number of iterations
        i += 1
      }
    }

    // Return the number of students that cannot eat lunch (the number of students remaining in the queue)
    queue.length
  }
}

File: dart/leet_code/1700.dart
/// Dart implementation of the LeetCode problem 1700: Number of Students Unable to Eat Lunch.
///
/// This module contains a class `Solution` with a method `countStudents`.
/// The `countStudents` method takes in two parameters:
/// - `students`: A list of integers representing the students' sandwich preferences.
/// - `sandwiches`: A list of integers representing the types of sandwiches.
///
/// The method returns an integer representing the number of students who cannot eat lunch.
/// The students are standing in a queue. Each student either prefers the type 0 sandwich or type 1 sandwich.
/// The sandwiches are stacked. Each sandwich is either type 0 or type 1. The students will eat in the following order:
/// - The student at the front of the queue will leave the queue and will eat the sandwich at the top of the stack only if it is the type that they prefer.
/// - Otherwise, they will leave it and wait at the end of the queue. The student will not eat at all if there is no sandwich of their preferred type.
///
/// Error Handling:
/// - This method assumes that the input parameters are well-formed, i.e., `students` and `sandwiches` are lists of integers.
/// - If the input parameters are not well-formed, the behavior of the method is undefined.

class Solution {
  /// Counts the number of students who cannot eat lunch.
  ///
  /// This method uses a while loop to simulate the process of the students eating the sandwiches.
  /// It iterates until the queue of students is empty. If the student at the front of the queue prefers the sandwich at the top of the stack, the student eats the sandwich and leaves the queue.
  /// Otherwise, the student waits at the end of the queue. If all students in the queue do not prefer the sandwich at the top of the stack, the loop breaks.
  /// After the loop, it returns the number of students who cannot eat lunch.
  ///
  /// @param students The students' sandwich preferences.
  /// @param sandwiches The types of sandwiches.
  /// @return The number of students who cannot eat lunch.
  int countStudents(List<int> students, List<int> sandwiches) {
    int count = 0;
    while (students.isNotEmpty) {
      if (students[0] == sandwiches[0]) {
        students.removeAt(0);
        sandwiches.removeAt(0);
        count = 0;
      } else {
        students.add(students[0]);
        students.removeAt(0);
        count++;
      }
      if (count == students.length) {
        break;
      }
    }
    return students.length;
  }
}
